docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1379,6 +1364,120 @@ As a Zephyr OS user I want to... ???
 
 [/SECTION]
 
+[SECTION]
+TITLE: Thread Synchronization
+
+[REQUIREMENT]
+UID: ZEP-990001
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Counting Semaphore Creation
+STATEMENT: >>>
+The system shall provide a mechanism to create semaphores, both statically at build time and dynamically at runtime, ensuring deterministic operation with bounded execution times.
+<<<
+USER_STORY: >>>
+As a software developer tasked with implementing resource management in our real-time operating system (RTOS), I want to integrate a semaphore synchronization primitive to facilitate coordinated access to shared resources among multiple tasks.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990002
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Initialization
+STATEMENT: >>>
+The system shall provide a mechanism to initialize semaphore values to regulate the number of tasks or resources concurrently accessing a shared resource.
+<<<
+RELATIONS:
+- TYPE: Parent
+  VALUE: ZEP-99
+
+[REQUIREMENT]
+UID: ZEP-990010
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Semaphore
+TITLE: Semaphore Initialization
+STATEMENT: >>>
+When initializing a counting semaphore, the system shall accept options specifying initial semaphore value and the maximum permitted count a semaphore can have.
```

Comment:
**question:** What about possible outcomes, success and failure? 

**thought:** To avoid redundancy Error management could be made a cross-cutting requirement for all Kernel API functions, e.g. to follow POSIX convention for error codes. Another such cross-cutting RQT would be the naming convention, e.g. common "k_" prefix and then kernel object specific identifier "sem_" for semaphore.